### ISSUE TITLE
`PwBaseWorkChain`: consider reinstating depending on "second consecutive unhandled failure" mechanism

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-quantumespresso#493]
The `BaseRestartWorkChain` comes with built in support to abort if a calculation fails twice in a row without any process handler having dealt with the problem. Sometime we stopped relying on this functionality because often if a calculation would fail due to an unexpected failure, such as an OOM error, restarting would be pointless, as the restart would be guaranteed to fail again. However, there are other unexpected failures that are more transient and might just resolve themselves with a simple restart. For example, recently I had hundreds of calculations fail on Daint due to node failures. Since the current `_handle_unrecoverable_failure` instantly aborts the workchain following the 300 of the `PwCalculation` the entire workchain upstream was also aborted. Had we simply retried one more time, we might have gotten lucky and would have saved not just this `PwBaseWorkChain` but the entire work chain in which it might have been nested.

So this comes down to a tradeoff:
Run the risk of pointlessly rerunning a failed calculation, or aborting a work chain due to a transient problem and causing the entire process stack to likely fail.

Currently we make the decision to better not run the risk of wasted resources by rerunning, but I am starting to wonder if this is the most efficient choice. Maybe the majority of these non-recoverable problems fail early and so the amount of potentially wasted resources is small. Especially considering all the potentially lost resources of all other sub processes successfully completed by the parent process, by aborting the work chain after the first try.

**Comment by sphuber** [comment]
@chrisjsewell , @giovannipizzi and @greschd would appreciate your gut feeling and input here

**Comment by greschd** [comment]
Yeah, I've definitely had cases where the error was "random", and simply restarting did the trick.

I think it might only make sense for the exit codes where `invalidates_cache` is `True` - since `invalidates_cache=False` kind of implies that the error is "real". I'm not sure we should mix this with the caching logic though - when caching is enabled the "retry" will effectively just be copying the same calculation anyway.

**Comment by giovannipizzi** [comment]
@sphuber Do we have some statistics on how often we incurred in errors that have been (or could have been) recovered with a simple restart, and how many times we have (or would have) rerun a calculation without reason, out of your big DB? And maybe the cost of this? Maybe this is a good exercise to go through for Marnik?

**Comment by sphuber** [comment]
It might indeed be a good exercise, but it would be an "advanced" one already, because it really requires to know to look even for errors that are not parsed yet. But once he is familiar with the benchmark project and I haven't dealt with this already we can ask him to look into it yes

**Comment by giovannipizzi** [comment]
In any case, pinging @mbercx with his GitHub handle so he's aware

**Comment by chrisjsewell** [comment]
> Run the risk of pointlessly rerunning a failed calculation, or aborting a work chain due to a transient problem and causing the entire process stack to likely fail.

Well, as a user, I would be in favour of the first option; giving the calculations every opportunity to succeed without my interjection. But then again, I've never had to be the one directly paying for these resources, so I'm not as worried about wasting them!
